rewrite-core/src/main/java/org/openrewrite/internal/ListUtils.java

```diff
@@ -146,50 +146,65 @@ public static <T> List<T> flatMap(@Nullable List<T> ls, BiFunction<Integer, T, O
             //noinspection ConstantConditions
             return ls;
         }
+
         List<T> newLs = ls;
-        int j = 0;
-        for (int i = 0; i < ls.size(); i++, j++) {
+        boolean widened = false;
+        List<T> outputLs = new ArrayList<>();
+
+        for (int i = 0; i < ls.size(); i++) {
             T tree = ls.get(i);
             Object newTreeOrTrees = flatMap.apply(i, tree);
-            if (newTreeOrTrees != tree) {
-                if (newLs == ls) {
-                    newLs = new ArrayList<>(ls);
-                }
-
+            if (newTreeOrTrees == tree) {
+                //noinspection unchecked
+                outputLs.add( (T) newTreeOrTrees);
+            } else {
                 if (newTreeOrTrees instanceof Iterable) {
-                    boolean addedTree = false;
                     //noinspection unchecked
-                    for (T newTree : (Iterable<T>) newTreeOrTrees) {
-                        if (j >= newLs.size()) {
-                            newLs.add(newTree);
-                        } else if (!addedTree) {
-                            newLs.set(j, newTree);
+                    Iterable<T> it = (Iterable<T>) newTreeOrTrees;
+                    List<T> outLs = new ArrayList<>();
```

Comment:
add a preceding if statement to see if it a `Collection` first, because we can initialize this list to the exact number of elements needed.